internal/carwriter/carwriter.go

```diff
@@ -0,0 +1,413 @@
+//
+// An error writing to a CAR file, or context concellation, stops processing
+// the advertisements from the datastore.
+func (cw *CarWriter) WriteExisting(ctx context.Context) <-chan int {
+	done := make(chan int, 1)
+
+	adCids, err := findAds(context.Background(), cw.dstore)
+	if err != nil {
+		log.Errorw("Error loading existing advertisements from datastore", "err", err)
+		close(done)
+		return done
+	}
+
+	if len(adCids) == 0 {
+		close(done)
+		return done
+	}
+
+	go func() {
+		log.Infof("Writing %d advertisements from datastore to CAR files", len(adCids))
+		var count int
+		for _, adCid := range adCids {
+			if ctx.Err() != nil {
+				break
+			}
+			_, _, err = cw.Write(ctx, adCid, false)
+			if err != nil {
+				log.Errorw("Cannot write advertisement to CAR file", "err", err)
+				var werr *WriteError
+				if errors.As(err, &werr) {
+					// Error writing to car file; stop writing ads.
+					break
+				}
+				// Log error, but keep going.
+				continue
+			}
+			count++
+		}
+		log.Infof("Wrote %d of %d advertisements from datastore to CAR files", count, len(adCids))
+		done <- count
+		close(done)
+	}()
+
+	return done
+}
+
+// Write reads the advertisement, specified by CID, from the datastore and
+// writes it and its entries into CAR file. The car file is stored in the
+// directory specified when the CarWriter was created, and is named with the
+// advertisement CID.
+//
+// The CAR file is written without entries if skipEntries is true. The prupose
+// of this to create a car file, to maintain the link in the advertisement
+// chain, when it is know that a later advertisement deletes this
+// advertisement's entries.
+func (cw *CarWriter) Write(ctx context.Context, adCid cid.Cid, skipEntries bool) (string, string, error) {
+	ad, data, err := cw.loadAd(adCid)
+	if err != nil {
+		return "", "", fmt.Errorf("cannot load advertisement: %w", err)
+	}
+
+	adCarFile, err := cw.writeAdToCar(ctx, adCid, data)
+	if err != nil {
+		return "", "", &WriteError{err}
+	}
+
+	if skipEntries || ad.Entries == schema.NoEntries || ad.Entries == nil {
+		cw.deleteCids([]cid.Cid{adCid})
+		log.Infow("Wrote advertisement CAR", "adFile", adCarFile.Path, "adFileSize", adCarFile.Size)
+		return adCarFile.Path, "", nil
+	}
+
+	entriesCid := ad.Entries.(cidlink.Link).Cid
+	if entriesCid == cid.Undef {
+		cw.deleteCids([]cid.Cid{adCid})
+		return adCarFile.Path, "", nil
+	}
+
+	entCarFile, err := cw.writeEntriesToCar(ctx, adCid, entriesCid)
+	if err != nil {
+		return "", "", err
+	}
+
+	log.Infow("Wrote advertisement CAR", "adFile", adCarFile.Path, "adFileSize", adCarFile.Size,
+		"entriesFile", entCarFile.Path, "entriesFileSize", entCarFile.Size)
+
+	return adCarFile.Path, entCarFile.Path, nil
+}
+
+// writeAdToCar wites the advertisement data to a CAR file.
+func (cw *CarWriter) writeAdToCar(ctx context.Context, adCid cid.Cid, data []byte) (*filestore.File, error) {
+	fileName := adCid.String() + "_adv.car"
```

Comment:
I am curious why include `_adv` and `_mhs` suffix? Would CID alone not be enough? 

If we add such prefixes, a client that only has a CID and does not know what node it contains would need to hit two endpoints.

If you think we need this, would it make sense to use path instead of file suffix? In the case of s3 that gives us advantages in terms of controlling things like retention per node type.

Reply:
All good points; agree.